### PR TITLE
Add a module for scalar conservation equations

### PR DESCRIPTION
**Pull request by tamiko**

### CODE REVIEW

**Comment by tamiko** [comment]
@Gerbeset *ping* FYI.

**Comment by ejtovar** [comment]
@tamiko Currently reviewing. Latest commit compiles fine, but 38 out of 45 tests failed... investigating now. Would you happen to have a prm file for the KPP problem or another standard test?

EDIT: The 38 failing tests were due to a bad cmake configuration. Only 1 test fails now. See other comment.

File: source/scalar_conservation/hyperbolic_system.h [hunk 1]
```diff
@@ -0,0 +1,901 @@
...
+       * And finally populate the flux list with all flux configurations
+       * defined in the FluxLibrary namespace:
+       */
+      FluxLibrary::populate_flux_list(flux_list_, subsection);
+
+      const auto populate_functions = [this]() {
+        bool initialized = false;
+        for (auto &it : flux_list_)
+
+          /* Populate EOS-specific quantities and functions */
```
Comment (ejtovar) [inline source/scalar_conservation/hyperbolic_system.h]: Change to `/* Populate flux functions */`

File: source/scalar_conservation/hyperbolic_system.h [hunk 2]
```diff
@@ -0,0 +1,901 @@
...
+            problem_name = "Scalar conservation equation (" + it->name() +
+                           ": " + it->flux_formula() + ")";
+            initialized = true;
+            break;
+          }
+
+        AssertThrow(
+            initialized,
+            dealii::ExcMessage(
+                "Could not find an equation of state description with name \"" +
```
Comment (ejtovar) [inline source/scalar_conservation/hyperbolic_system.h]: Change to `"Could not find a flux function description with name \""` ...

File: source/scalar_conservation/hyperbolic_system.h [hunk 3]
```diff
@@ -0,0 +1,901 @@
...
+      riemann_solver_random_entropies_ = 0;
+      add_parameter(
+          "riemann solver random entropies",
+          riemann_solver_random_entropies_,
+          "In addition to the wavespeed estimate based on the Roe average and "
+          "flux gradients of the left and right state also enforce an entropy "
+          "inequality on the prescribed number of random Krŭzkov entropies.");
+
+      riemann_solver_approximation_delta_ = 1.0e-10;
+      add_parameter("riemann solver approximation delta",
```
Comment (ejtovar) [inline source/scalar_conservation/hyperbolic_system.h]: I need to refresh on the scalar conservation law theory, but why does this need to be a user parameter?

**Review by ejtovar — CHANGES_REQUESTED** [review]
@tamiko This looks great! Thanks for this. I added some comments on a few lines. And two things:

1. Do you mind adding a default `.prm` file for the KPP problem?

2. The riemann solver tests for the scalar conversation law module fail on my computer in a funny way. The first two letters of the flux description for the tests in 2D:

```
test<2, double>("burgers");

test<2, double>("kpp");
```

get cutoff when reading in. For example:

```
An error occurred in line <619> of file </Users/ericjtovar/Software/ryujin/source/scalar_conservation/hyperbolic_system.h> in function
    auto ryujin::ScalarConservationTest::HyperbolicSystem::HyperbolicSystem(const std::string &)::(anonymous class)::operator()() const
The violated condition was:
    initialized
Additional information:
    Could not find an equation of state description with name "rgers"
```

These are some quick comments for now. Will continue to review thoroughly.

**Comment by ejtovar** [comment]
@tamiko Should I fix small typos or just point them out here?

File: source/euler/initial_state_function.h [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+#include <initial_state_library.h>
+
+#include <deal.II/base/function_parser.h>
+
+namespace ryujin
+{
+  namespace EulerInitialStates
+  {
+    /**
+     * Returns a initial state defined by a set of user specified function.
```
Comment (ejtovar) [inline source/euler/initial_state_function.h]: `* Returns an initial state...`

**Comment by tamiko** [comment]
@ejtovar Thanks for the review. I will take care of the typos and the other comments.

**Comment by tamiko** [comment]
@ejtovar
1. I plan to add a lot of benchmark and example prms over the coming weeks (not just for scalar conservation equations, but for all conservation law modules we have implemented). For now, I would like to merge the core functionality with this pull request.
2. This is a baffling test failure - in particular all tests pass on my machine and on the CI (`verification/tests-deal.II-master`). Would you mind to open an issue for this test failure so that we can investigate what is going wrong? In particular, would you mind sending me the output of `otool -L ./ryujin`? I want to make sure that you don't accidentally link against both `libc++` (Clang's C++ standard library) and `libstdc++` (GCC's C++ standard library).

File: source/scalar_conservation/hyperbolic_system.h (hunk 3 above)
Comment (tamiko) [reply source/scalar_conservation/hyperbolic_system.h]: I will refactor this into an internally chosen parameter.

**Comment by ejtovar** [comment]
@tamiko 

1. I figured, but I just wanted to test the module myself with standard problems and verify with the "eye norm" :-)

2. Here is output of the otool command. 

```
ryujin:
	/Users/ericjtovar/Software/dealii-install/lib/libdeal_II.9.6.0-pre.dylib (compatibility version 9.6.0, current version 9.6.0)
	/opt/homebrew/opt/open-mpi/lib/libmpi_usempif08.40.dylib (compatibility version 71.0.0, current version 71.0.0)
	/opt/homebrew/opt/open-mpi/lib/libmpi_usempi_ignore_tkr.40.dylib (compatibility version 71.0.0, current version 71.0.0)
	/opt/homebrew/opt/open-mpi/lib/libmpi_mpifh.40.dylib (compatibility version 71.0.0, current version 71.0.0)
	/opt/homebrew/opt/tbb/lib/libtbb.12.dylib (compatibility version 12.0.0, current version 12.10.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/opt/homebrew/opt/boost/lib/libboost_iostreams-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/homebrew/opt/boost/lib/libboost_serialization-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/homebrew/opt/boost/lib/libboost_system-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/homebrew/opt/boost/lib/libboost_thread-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/homebrew/opt/boost/lib/libboost_regex-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/homebrew/opt/boost/lib/libboost_chrono-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/homebrew/opt/boost/lib/libboost_atomic-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/homebrew/opt/assimp/lib/libassimp.5.dylib (compatibility version 5.0.0, current version 5.2.4)
	/opt/homebrew/opt/gmp/lib/libgmpxx.4.dylib (compatibility version 11.0.0, current version 11.1.0)
	/opt/homebrew/opt/mpfr/lib/libmpfr.6.dylib (compatibility version 9.0.0, current version 9.0.0)
	/opt/homebrew/opt/gmp/lib/libgmp.10.dylib (compatibility version 15.0.0, current version 15.1.0)
	/opt/homebrew/opt/gmsh/lib/libgmsh.4.11.dylib (compatibility version 4.11.0, current version 4.11.1)
	/opt/homebrew/opt/opencascade/lib/libTKBO.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKBool.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKBRep.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKernel.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKFeat.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKFillet.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKG2d.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKG3d.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKGeomAlgo.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKGeomBase.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKHLR.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKIGES.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKMath.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKMesh.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKOffset.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKPrim.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKShHealing.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKSTEP.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKSTEPAttr.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKSTEPBase.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKSTEP209.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKSTL.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKTopAlgo.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/opt/homebrew/opt/opencascade/lib/libTKXSBase.7.7.dylib (compatibility version 7.7.0, current version 7.7.2)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
	/System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate (compatibility version 1.0.0, current version 4.0.0)
	/opt/homebrew/opt/open-mpi/lib/libmpi.40.dylib (compatibility version 71.0.0, current version 71.5.0)
	/Users/ericjtovar/Software/p4est-install/FAST/lib/libp4est.2.dylib (compatibility version 3.0.0, current version 3.0.0)
	/Users/ericjtovar/Software/p4est-install/FAST/lib/libsc.2.dylib (compatibility version 3.0.0, current version 3.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1300.36.0)
```

EDIT (09/01): realized I used the wrong flag for otool, updated now

**Comment by tamiko** [comment]
@ejtovar https://gist.github.com/tamiko/3fc331979224de127793355d7cebf09a

**Comment by ejtovar** [comment]
@tamiko Still having the weird string cutoff issue. I will investigate shortly.

**Comment by tamiko** [comment]
@ejtovar This is now fixed.

**Review by ejtovar — APPROVED** [review]
Everything is good to go!

**Pull request merged**